pkg/controllers/managementuser/resourcequota/resource_quota_sync.go

```diff
@@ -119,68 +117,65 @@ func (c *SyncController) CreateResourceQuota(ns *corev1.Namespace) (runtime.Obje
 		return ns, err
 	}
 
-	projectLimit, _, err := getProjectResourceQuotaLimit(ns, c.ProjectLister)
-	if err != nil {
-		return ns, err
-	}
-
-	var quotaSpec *corev1.ResourceQuotaSpec
-	if projectLimit != nil {
-		quotaSpec, err = c.getNamespaceResourceQuota(ns)
-		if err != nil {
-			return ns, err
-		}
-	}
-
-	quotaToUpdate, err := c.getResourceQuotaToUpdate(ns)
+	requestedQuotaLimit, newQuotaSpec, err := c.deriveRequestedResourceQuota(ns)
 	if err != nil {
 		return ns, err
 	}
 
 	operation := "none"
 	if existing == nil {
-		if quotaSpec != nil {
+		if newQuotaSpec != nil {
 			operation = "create"
 		}
 	} else {
-		if quotaSpec == nil {
+		if newQuotaSpec == nil {
 			operation = "delete"
-		} else if quotaToUpdate != "" || !apiequality.Semantic.DeepEqual(existing.Spec.Hard, quotaSpec.Hard) {
+		} else if !apiequality.Semantic.DeepEqual(existing.Spec.Hard, newQuotaSpec.Hard) {
 			operation = "update"
 		}
 	}
 
 	var updated *corev1.Namespace
-	var isFit bool
+	var operationErr error
 	switch operation {
 	case "create":
-		isFit, updated, err = c.validateAndSetNamespaceQuota(ns, quotaToUpdate)
+		isFit, updated, exceeded, err := c.validateAndSetNamespaceQuota(ns, &v32.NamespaceResourceQuota{Limit: *requestedQuotaLimit})
 		if err != nil {
 			return updated, err
 		}
 		if !isFit {
-			// create default "all 0" resource quota
-			quotaSpec, err = getDefaultQuotaSpec()
+			// Create a quota with zeros only for overused resources.
+			limit, err := zeroOutResourceQuotaLimit(requestedQuotaLimit, exceeded)
+			if err != nil {
+				return updated, err
+			}
+
+			newQuotaSpec, err = convertResourceLimitResourceQuotaSpec(limit)
 			if err != nil {
 				return updated, err
 			}
 		}
-		err = c.createDefaultResourceQuota(ns, quotaSpec)
+		operationErr = c.createResourceQuota(ns, newQuotaSpec)
 	case "update":
-		isFit, updated, err = c.validateAndSetNamespaceQuota(ns, quotaToUpdate)
-		if err != nil || !isFit {
+		isFit, updated, exceeded, err := c.validateAndSetNamespaceQuota(ns, &v32.NamespaceResourceQuota{Limit: *requestedQuotaLimit})
+		if err != nil {
 			return updated, err
 		}
-		err = c.updateResourceQuota(existing, quotaSpec)
+		if !isFit {
+			operationErr = fmt.Errorf("update quota doesn't fit due to the following resource requests: [%v]", format.ResourceList(exceeded))
```

Comment:
This seems like a behavior change that could have some impact. Before, if `isFit` was false on update, then this would return nil and appear successful. Are you sure that this new behavior, where a bad fit now returns an error, is correct behavior? (It would seem to make sense to me but I want to be sure the ramifications have been thought through.)

Reply:
Good point. The difference is that now we won't execute the last statements of the method, which set informational annotations on the namespace, like the whether validation of the namespace failed and, if yes, why. But we want that.

Reply:
Pushed an update so that the annotation is always written to a namespace - now on update the exact exceeded resources are shown.